### ISSUE TITLE
UTF-8 Strings are parsed as latin1.

### ISSUE THREAD

**Issue by Ivorforce** [godotengine/godot#100641]
### Tested versions

All versions are affected. This is _old_ code.

### System information

N/A

### Issue description

The current implementation of Godot has a bug where UTF-8 encoded strings are parsed as `latin1`.

I'd like to hear what you all think is the best solution to this problem.

## About the Issue

The `String(const char *)` constructor parses using latin1 encoding:

https://github.com/godotengine/godot/blob/89001f91d21ebd08b59756841426f540b154b97d/core/string/ustring.h#L615-L617

(`parse_latin1` was [recently renamed](https://github.com/godotengine/godot/pull/100434) from `copy_from` because that's what the encoding is interpreted as).

This constructor is used for many strings, including static C strings. The encoding of static C strings is controlled by `-fexec-charset` on GCC / Clang[^1] and `/execution-charset` on MSVC[^2]. In the Godot codebase, it defaults to UTF-8 on GCC / Clang, and is [explicitly set](https://github.com/godotengine/godot/blob/89001f91d21ebd08b59756841426f540b154b97d/platform/windows/detect.py#L448) to UTF-8 on Windows. Therefore, there is a mismatch between encodings. It is likely that there are mismatches in other, non-static use-cases. 

The mismatch has apparently [led to problems](https://github.com/godotengine/godot/pull/100434#discussion_r1885808663) a few times in the past (though I don't know which). For the regular case, strings use ascii-only characters, in which range `latin1` and `utf-8` overlap.

## Possible Solutions

- Enforce ASCII-only for static C strings (and force the use of `u8"String"` for utf-8 strings).
    - Using `-fexec-charset` (and force the use of `u8"String"` for utf-8 strings). This would be the best solution in my opinion, but it doesn't appear to be possible to pass anything else except `UTF-8` to `-fexec-charset`.
    - Using external linter tools or options. I don't know if such a tool exists.
- Use UTF-8 parsing for strings for the default constructors. This would be somewhat slower than `latin1` parsing (though accelerated by https://github.com/godotengine/godot/pull/99826#pullrequestreview-2504675624). The exact difference would have to be measured and tested for significance.
- Use ASCII-only parsing for strings for the default constructors. Log errors if values > 127 are encountered. This should be negligibly slower than `parse_latin1`, and somewhat faster than `parse_utf8`.

Ideally, the default, implicit `String` constructor is removed, and every use of it replaced with an explicit call to `String::utf8` or similar (maybe with the exception of construction from string literals), so we can be sure intent was put behind choosing the encoding. This is a bigger task though and not realistic for the near future.

[^1]: https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html#index-fexec-charset
[^2]: https://learn.microsoft.com/en-us/cpp/build/reference/execution-charset-set-execution-character-set?view=msvc-170

### Steps to reproduce

N/A

### Minimal reproduction project (MRP)

N/A

**Comment by AThousandShips** [comment]
The choice to make `String(char *)` only accept Latin1 is by design as a performance benefit, but needs to be communicated better to avoid issues

**Comment by Ivorforce** [comment]
> The choice to make `String(char *)` only accept Latin1 is by design as a performance benefit, but needs to be communicated better to avoid issues

The actual performance gain for this decision should be tested. Correctness is always more important than performance. I don't think switching to `ascii` (or perhaps even `utf8`) would introduce that much of a detriment.

**Comment by AThousandShips** [comment]
I agree, there was talk of adding a check for correctness as well, but *almost all* string literals are latin1 so that's important to keep in mind as well, or rather ASCII

**Comment by AThousandShips** [comment]
Agreed, handling things with literals specially would be very useful, most if not all dynamic input is processed as utf8 already, with a few exceptions that should probably use the explicit case of `ascii` instead to make things explicit

**Comment by kiroxas** [comment]
Does anyone has some data on this ? Like how many static strings are parsed in the codebase, and what's the average size for them. Because if we're talking 20 bytes string, the difference between thise should be very small. I'm not even sure parse utf8 parsing only ascii would ba that slow. But we can probably measure that. I have not access to my computer for some time, but it would be nice to gather those metrics, to avoid guessing.

**Comment by bruvzg** [comment]
Not sure if there's any data, but you can estimate amount of string literals by checking translation files (`editor/translations`), it won't include console only error messages and some other stuff, but should give a lower bound. And since engine UI translation is done via `po` files, all string literals in the code in general are in English (ASCII only), there are only few special cases using non ASCII characters (`µs`, `•`, `⚠`, `×`, `±`, box drawing characters, font preview sample strings, few special strings for pseudolocalizations).

**Comment by kiroxas** [comment]
I think we can sefely ignore all files from `po` files, as they uses `parse_utf8` already anyway, this would not change anything for them.

I did some quick benchmarking for the cost of `parse_utf8` vs `parse_latin`.
I tested 2 kind of strings, long ones around (500 bytes) to amortize the startup cost (aka memory allocation), and short ones (around 20 bytes), probably more relevant to static string in the codebase. (parse_utf8_PR is the implementation from [here](https://github.com/godotengine/godot/pull/99826))

| Function | Cycles (Long) | Cycles (Short)|
| :---------------- | :------: | :----: |
| `latin1`|   5 | 39 |
| `parse_utf8`           |   11 | 60 |
| `parse_utf8_PR`    |  8.5| 55 |

So in the case of small strings, we're talking about 200-300 cycles difference per string. This is not nothing, but if you happen to do it 50 times in a frame, this is a 10k cycles budget. 
The right question now is, is this cost too much to remove some unlikely errors ? (might be, might not be).
If those static strings are not transformed into Strings every frame, but only once on startup, then I think we should parse_utf8 by default, as the difference would not be noticeable, even for thousands of those. 

If anyone wants to test this, here is the code, to drop in `Main::iteration` somewhere ( maybe need some tuning for the rdtsc intrinsics depending on your compiler)
```
const char mySuperString[] = "Does anyone has some ?";

static uint64_t latin = 0;
static uint64_t utf = 0;
static uint64_t iterations = 0;

++iterations;
uint64_t start = __rdtsc();
String normalString(mySuperString);
uint64_t end = __rdtsc();
String utfString;
utfString.parse_utf8(mySuperString);
uint64_t utfEnd = __rdtsc();

latin += end - start;
utf += utfEnd - end;

real_t latins = (real_t)(latin / iterations);
real_t utfs   = (real_t)(utf / iterations);
uint32_t size = sizeof(mySuperString) / sizeof(mySuperString[0]);

print_line(vformat("(%d:%d)  Normal : %f : %f, Utf: %f : %f", iterations, size, latins, latins / size, utfs,  utfs / size));
```
